torch/testing/_internal/common_utils.py

```diff
@@ -1309,30 +1336,148 @@ def check_if_enable(test: unittest.TestCase):
+                isinstance(expected, self._supported_types)
+                and isinstance(actual, (*self._supported_types, torch.Tensor, np.ndarray))
+            )
+        ):
+            raise UnsupportedInputs()
+
+        return [self._to_number(input, id=id) for input in (actual, expected)]
+
+    def _to_number(self, number_like, *, id):
+        if isinstance(number_like, (torch.Tensor, np.ndarray)):
+            numel = number_like.numel() if isinstance(number_like, torch.Tensor) else number_like.size
+            if numel > 1:
+                raise ErrorMeta(
+                    ValueError,
+                    f"Only single element tensor-likes can be compared against a number. "
+                    f"Got {numel} elements instead.",
+                    id=id,
+                )
+            number = number_like.item()
+            if isinstance(number, bool):
+                number = int(number)
+
+            return number
+        # We need this check, because combinations of an int and a bool will not be picked up by the BooleanPair,
+        # but will be picked up here since issubclass(bool, int)
+        elif isinstance(number_like, bool):
+            return int(number_like)
+        elif isinstance(number_like, Enum):
+            return int(number_like)  # type: ignore[call-overload]
+        else:
+            return super()._to_number(number_like, id=id)
+
+
+class TensorOrArrayPair(TensorLikePair):
+    def __init__(self, actual, expected, *, rtol_override=0.0, atol_override=0.0, **other_parameters):
+        super().__init__(actual, expected, **other_parameters)
+        self.rtol = max(self.rtol, rtol_override)
+        self.atol = max(self.atol, atol_override)
+
+    def _process_inputs(self, actual, expected, *, id, allow_subclasses):
+        self._check_inputs_isinstance(actual, expected, cls=(torch.Tensor, np.ndarray))
+
+        actual, expected = [self._to_tensor(input, id=id) for input in (actual, expected)]
+        for tensor in (actual, expected):
+            self._check_supported(tensor, id=id)
+        return actual, expected
+
+
+class UnittestPair(Pair):
+    CLS: Union[Type, Tuple[Type, ...]]
+    TYPE_NAME: Optional[str] = None
+
+    def __init__(self, actual, expected, **other_parameters):
+        self._check_inputs_isinstance(actual, expected, cls=self.CLS)
+        super().__init__(actual, expected, **other_parameters)
+
+    def compare(self):
+        test_case = unittest.TestCase()
+
+        try:
+            return test_case.assertEqual(self.actual, self.expected)
+        except test_case.failureException as error:
+            msg = str(error)
+
+        type_name = self.TYPE_NAME or (self.CLS if isinstance(self.CLS, type) else self.CLS[0]).__name__
+        return self._make_error_meta(AssertionError, f"{type_name.title()} comparison failed: {msg}")
+
+
+class StringPair(UnittestPair):
+    CLS = string_classes
+    TYPE_NAME = "string"
+
+
+class SetPair(UnittestPair):
```

Comment:
Do we not handle sets in assert_close?

Reply:
The old implementation had support for it:

https://github.com/pytorch/pytorch/blob/e3bcf64ff84f8e96839e39056b3b90d1bd1f8bbe/torch/testing/_internal/common_utils.py#L2025-L2028

 I'll remove it temporary and see how many tests break.

Reply:
Oh sorry, I think I misunderstood: no I don't think we should support sets in `assert_close`, because that would mean we need to somehow match tensors in two sets before we can do an actual numeric comparison. That is easy for sequences and mappings by index / key, but I don't think that is possible for unordered containers.